File: etl/src/main/scala/org/apache/spark/metrics/source/StreamingSource.scala
package org.apache.spark.metrics.source

import com.codahale.metrics.{Gauge, MetricRegistry}
import org.apache.spark.sql.streaming.StreamingQueryProgress

object StreamingSource {


}

class StreamingSource(val appName: String) extends Source {
  register("batchId", () => _progress.batchId)
  register("numInputRows", () => _progress.numInputRows)
  register("inputRowsPerSecond", () => _progress.inputRowsPerSecond)
  register("processedRowsPerSecond", () => _progress.processedRowsPerSecond)
  final val _metricRegistry = new MetricRegistry
  final val _progress = new Progress

  private def register[T](name: String, metric: Gauge[T]) = _metricRegistry.register(MetricRegistry.name(name), metric)

  override def sourceName: String = String.format("%s.streaming", appName)

  override def metricRegistry: MetricRegistry = _metricRegistry

  def updateProgress(queryProgress: StreamingQueryProgress): Unit = {
    _progress.batchId(queryProgress.batchId)
    _progress.numInputRows(queryProgress.numInputRows)
    _progress.inputRowsPerSecond(queryProgress.inputRowsPerSecond)
    _progress.processedRowsPerSecond(queryProgress.processedRowsPerSecond)
  }

  class Progress {
    private var _batchId: Long = -1
    private var _numInputRows: Long = 0
    private var _inputRowsPerSecond: Double = 0
    private var _processedRowsPerSecond: Double = 0


    def batchId: Long = _batchId

    def batchId(batchId: Long): Unit = {
      _batchId = batchId
    }

    def numInputRows: Long = _numInputRows

    def numInputRows(numInputRows: Long): Unit = {
      _numInputRows = numInputRows
    }

    def inputRowsPerSecond: Double = _inputRowsPerSecond

    def inputRowsPerSecond(inputRowsPerSecond: Double): Unit = {
      _inputRowsPerSecond = inputRowsPerSecond
    }

    def processedRowsPerSecond: Double = _processedRowsPerSecond

    def processedRowsPerSecond(processedRowsPerSecond: Double): Unit = {
      _processedRowsPerSecond = processedRowsPerSecond
    }
  }
}

File: etl/src/main/scala/org/apache/spark/metrics/source/bla/StreamingSource.java
//package org.apache.spark.metrics.source.bla;
//
//import com.codahale.metrics.Gauge;
//import com.codahale.metrics.MetricRegistry;
//import org.apache.spark.metrics.source.Source;
//import org.apache.spark.sql.streaming.StreamingQueryProgress;
//
//
//public class StreamingSource implements Source {
//    private final String appName;
//    private final MetricRegistry metricRegistry = new MetricRegistry();
//    private final Progress progress = new Progress();
//
//    public StreamingSource(String appName) {
//        this.appName = appName;
//        register("batchId", progress::getBatchId());
//        register("numInputRows", progress::getNumInputRows);
//        register("inputRowsPerSecond", progress::getInputRowsPerSecond);
//        register("processedRowsPerSecond", progress::getProcessedRowsPerSecond);
//    }
//
//    private void register(String name, Gauge metric) {
//        metricRegistry.register(MetricRegistry.name(name), metric);
//    }
//
//    @Override
//    public String sourceName() {
//        return String.format("%s.streaming", appName);
//    }
//
//
//    @Override
//    public MetricRegistry metricRegistry() {
//        return metricRegistry;
//    }
//
//    public void updateProgress(StreamingQueryProgress queryProgress) {
//        progress.setBatchId(queryProgress.batchId());
//        progress.setNumInputRows(queryProgress.numInputRows());
//        progress.setInputRowsPerSecond(queryProgress.inputRowsPerSecond());
//        progress.setProcessedRowsPerSecond(queryProgress.processedRowsPerSecond());
//    }
//
//    private static class Progress {
//
//        private long batchId = -1;
//        private long numInputRows = 0;
//        private double inputRowsPerSecond = 0;
//        private double processedRowsPerSecond = 0;
//
//        public long getBatchId() {
//            return batchId;
//        }
//
//        public void setBatchId(long batchId) {
//            this.batchId = batchId;
//        }
//
//        public long getNumInputRows() {
//            return numInputRows;
//        }
//
//        public void setNumInputRows(long numInputRows) {
//            this.numInputRows = numInputRows;
//        }
//
//        public double getInputRowsPerSecond() {
//            return inputRowsPerSecond;
//        }
//
//        public void setInputRowsPerSecond(double inputRowsPerSecond) {
//            this.inputRowsPerSecond = inputRowsPerSecond;
//        }
//
//        public double getProcessedRowsPerSecond() {
//            return processedRowsPerSecond;
//        }
//
//        public void setProcessedRowsPerSecond(double processedRowsPerSecond) {
//            this.processedRowsPerSecond = processedRowsPerSecond;
//        }
//    }
//}
